### PR TITLE
8329816: Add SLEEF version 3.6.1

### PR DESCRIPTION
**Pull request by vidmik**
[JDK-8312425](https://bugs.openjdk.org/browse/JDK-8312425) is looking to optimize vector math operations by leveraging the SLEEF library. For legal reasons the actual contribution of the SLEEF files needs to be handled separately. This enhancement adds the relevant files, enabling the rest of [JDK-8312425](https://bugs.openjdk.org/browse/JDK-8312425) to move forward.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329816](https://bugs.openjdk.org/browse/JDK-8329816): Add SLEEF version 3.6.1 (**Enhancement** - P4) ⚠️ Issue is already resolved. Consider making this a "backport pull request" by setting the PR title to `Backport <hash>` with the hash of the original commit. See [Backports](https://wiki.openjdk.org/display/SKARA/Backports).

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) Review applies to [0c91d4b1](https://git.openjdk.org/jdk/pull/19185/files/0c91d4b12e7bf4b0d179151877f4ec5961ec8130)
 * [Hamlin Li](https://openjdk.org/census#mli) (@Hamlin-Li - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19185/head:pull/19185` \
`$ git checkout pull/19185`

Update a local copy of the PR: \
`$ git checkout pull/19185` \
`$ git pull https://git.openjdk.org/jdk.git pull/19185/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19185`

View PR using the GUI difftool: \
`$ git pr show -t 19185`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19185.diff">https://git.openjdk.org/jdk/pull/19185.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19185#issuecomment-2105460642)

### CODE REVIEW

**Comment by Hamlin-Li** [comment]
Thanks for working on this.
Just one question, currently in `sleefinline_{advsimd,sve}.h` there is `#define SLEEF_ALWAYS_INLINE inline`, but what expected is something like `#define SLEEF_ALWAYS_INLINE inline __attribute__((always_inline))` (please check https://github.com/shibatch/sleef/pull/537 for details). In the future, when we change it, do we need to go through the legal process again? I guess we don't need it, but just double check it.

**Comment by vidmik** [comment]
Looks like that change is not yet in a released version of SLEEF, and in particular not in 3.6.

We do need updated approvals when we pick up new versions. Since we've gone through the process once it's typically easier/faster to do so. It will be technically easier/faster as well, now that I know how to do it and have encoded it in the createSleef.sh script.

File: src/jdk.incubator.vector/linux/legal/sleef.md
Comment (erikj79) [inline src/jdk.incubator.vector/linux/legal/sleef.md:32]: Is this missing line break a mistake or intended?

**Review by erikj79 — APPROVED** [review]

**Comment by Hamlin-Li** [comment]
> Looks like that change is not yet in a released version of SLEEF, and in particular not in 3.6.
> 
> We do need updated approvals when we pick up new versions. Since we've gone through the process once it's typically easier/faster to do so. It will be technically easier/faster as well, now that I know how to do it and have encoded it in the createSleef.sh script.

Thanks, I see.
As the version 3.6 will introduce some performance regression compared to non-intrinsic version, so to bring the fix into jdk, we need either do a manual change (like https://github.com/openjdk/jdk/pull/18605/commits/cd70f5a970514938d05f7a2426b15f78245236d3), or wait the next version after 3.6 (which should include changes in https://github.com/shibatch/sleef/pull/537), I think we prefer the latter (i.e. depends on next version after 3.6)
That means https://github.com/openjdk/jdk/pull/18605 (JDK-8312425) requires sleef version next to 3.6 (could be 3.6.1 or 3.7).

I'm OK to push this pr first, if it's also convenient for you, as we will need to push the change between 3.6.1/3.7 and 3.6 again when 3.6.1/3.7 is released.

**Review by Hamlin-Li — APPROVED** [review]

**Comment by vidmik** [comment]
Let me suggest that we wait for the next release of SLEEF to be released in that case since that release seems to be imminent. That way we'll get both the performance fixes *and* we can include the RISC-V header files at the same time. Fair?

**Comment by Hamlin-Li** [comment]
> Let me suggest that we wait for the next release of SLEEF to be released in that case since that release seems to be imminent. That way we'll get both the performance fixes _and_ we can include the RISC-V header files at the same time. Fair?

Yes, it makes sense to me. Thanks!

**Comment by vidmik** [comment]
Thank you Hamlin. I'll try to keep my eyes open for the announcement of the upcoming SLEEF release but do feel free to notify me if you see it first!

**Comment by Hamlin-Li** [comment]
> Thank you Hamlin. I'll try to keep my eyes open for the announcement of the upcoming SLEEF release but do feel free to notify me if you see it first!

Thank you @vidmik , sure I will do it.

File: make/devkit/createSleef.sh
Comment (magicus) [inline make/devkit/createSleef.sh:32]: So you'd need a different copy of sleef for each platform? The files you have put in `linux/native/libvectormath`, what platform are they for? Should we not put them in a platform-specific subdirectory?

Comment (Hamlin-Li) [reply make/devkit/createSleef.sh:32]: > So you'd need a different copy of sleef for each platform?

I think it's one or more.

> The files you have put in linux/native/libvectormath, what platform are they for? Should we not put them in a platform-specific subdirectory?

we could, but not necessary, as long as they have different suffixes, and normally that suffixes indicate what platform it's for.

Comment (magicus) [reply make/devkit/createSleef.sh:32]: Okay, suffix works fine too. But the files currently in the patch is named e.g. `sleefinline_advsimd.h`, which does not indicate any platform at all. Is it a generic file, and the platform specific ones are still missing from this PR?

Comment (Hamlin-Li) [reply make/devkit/createSleef.sh:32]: I think both `sleefinline_advsimd.h` and `sleefinline_sve.h` are specific for arm.
In the future, on riscv the corresponding file name will be `sleefinline_rvvm1.h`.

Comment (magicus) [reply make/devkit/createSleef.sh:32]: Oh, you mean that the `sve` suffix signals that it is for ARM. I thought you were talking about having a name like `sleefinline_aarch64.h`.

Sure, if the only files generated by sleef are ever .h files, the logic for chosing which to include can be done entirely by `#ifdef`s in the source code. But if there ever needs to be different .c or .cpp files to include in the build, the build system needs to be able to determine automatically if they should be included or included, and that can only be made if the path or the file name includes the CPU moniker. 

Personally, I think it would show good alignment with the prevailing norms in the JDK to also use this way of naming files for .h files. But I confess that for .h files it is more a matter of style, rather than a necessity from the build system.

Comment (Hamlin-Li) [reply make/devkit/createSleef.sh:32]: Thanks for the information. I think the files from sleef will be .h headers only. I'm fine to align with prevailing norms in the JDK, it's always good to do so.

Just a reminder for us in the future discussion of https://github.com/openjdk/jdk/pull/18605, maybe we should consider this dir or file naming norms, as there will be more files related to different platforms added in the future, e.g. riscv64.

**Comment by vidmik** [comment]
I, too, envision that we'll be importing header files (only). That said, I'd very much prefer *not* to rename them as part of the import. If anything I could see us have architecture specific directories in which we place the respective files (and a common directory for `misc.h`), but it's not immediately clear to me that it's worth it given that the files are used in such a narrow context in the JDK.

**Comment by magicus** [comment]
I understand that you want to avoid renaming files, if they are imported. That is a good point. Moving them to an arch subdirectory does not seem like much additional hassle (there's apparently still a lot of manual work involved in upgrading the source from the third party upstream), and might help readers that are not deeply familiar with these platforms. But then again, if we only talk about header files, it is not strictly needed, so if you don't want to do it, then skip it.

**Comment by Hamlin-Li** [comment]
In case you need it and avoid to generate yourself, I've generated sleef inline header of 3.6.1 for riscv, it's at https://github.com/openjdk/jdk/pull/18605/commits/c279a3c290554892939267d9ebe88198988d31a4

**Comment by vidmik** [comment]
@Hamlin-Li I have generated the header files (two aarch64 and the new one for riscv64) using SLEEF 3.6.1. Please make sure they match your expectations!

**Comment by Hamlin-Li** [comment]
@vidmik Thanks for updating, I think I'd better to verify it in case we need uncessary further changes. Will update when I finish.

**Comment by Hamlin-Li** [comment]
@vidmik I have verified the tests and performance for aarch64, it's good as before, I also updated related information in https://github.com/openjdk/jdk/pull/18605.
Although have some issue for riscv64, but I think it's some env issues rather than issue in `sleefinline_rvvm1.h`.
So, I think we are good to go with this pr.
Thanks a lot for the work!

File: src/jdk.incubator.vector/linux/native/libvectormath/README
Comment (Hamlin-Li) [inline src/jdk.incubator.vector/linux/native/libvectormath/README:18]: `-DSLEEF_ENFORCE_SVE=ON` can be removed.

File: src/jdk.incubator.vector/linux/native/libvectormath/README
Comment (Hamlin-Li) [inline src/jdk.incubator.vector/linux/native/libvectormath/README:41]: These lines can be removed now.

**Review by Hamlin-Li — APPROVED** [review]
Just some minor comments in README, otherwise looks good.

**Comment by Hamlin-Li** [comment]
An quick update, I just verified via QEMU that sleefinline_rvvm1.h works well too.

**Comment by Hamlin-Li** [comment]
Hi @vidmik @theRealAph ,
Just a kindly reminder, could you please bring up the topic (about integrating 3rd party source into jdk, mentioned in [1](https://github.com/openjdk/jdk/pull/18605#issuecomment-2214663777), [2](https://github.com/openjdk/jdk/pull/18605#issuecomment-2230581736)) on workshop or summit? Thanks!

**Comment by erikj79** [comment]
Responding to the discussion in https://github.com/openjdk/jdk/pull/18605 here as this is the PR actually adding SLEEF to the JDK source.

I was initially of the opinion that the solution already provided here was enough. We could potentially have added a Git hash in addition to the version/tag to more precisely and permanently identify the exact Sleef source we are depending on. At least a Git hash wouldn't change externally.

However, Andrew's arguments have swayed me. I now think we should add a more or less complete dump of the Sleef source into the JDK repository. I'm still open to trimming it down somewhat as long as the build steps we need to run to generate the headers we need, using the Sleef build system, are still functional. I'm basically agreeing with his suggestion but will spell it out in detail here for completeness in this PR.

We should then add a script that automatically performs the necessary build steps, using the Sleef official build system, to generate the headers we need, and outputs them into the JDK source tree, in the location where we will also commit those headers. This script should document what dependencies and configuration is necessary to run it, which will include cmake and possibly other things. Performing this step doesn't need to be completely streamlined, just reasonably possible to run. It's meant to be an import/verification step. With this solution I would recommend putting the script next to the Sleef source tree instead of in make/devkit/.

The normal JDK build will just use the committed pre-generated headers.

I looked briefly at the heroic build work provided by @fitzsim and while I admire the effort, I don't think this is the right way and we already dismissed this approach earlier. Not because I didn't think it was feasible to implement, but because of the future maintenance burden. The Sleef build is non trivial so we shouldn't try to replicate it in our build. The risk of our implementation diverging in the future is too great.

**Comment by Hamlin-Li** [comment]
> With this solution I would recommend putting the script next to the Sleef source tree instead of in make/devkit/.

> The normal JDK build will just use the committed pre-generated headers.

> Not because I didn't think it was feasible to implement, but because of the future maintenance burden.

I agree with @erikj79, in particular the above 3 points.

Shall we be ready to move forward? Or we're still blocked by some other issues, such as legal process?

**Comment by theRealAph** [comment]
> Responding to the discussion in #18605 here as this is the PR actually adding SLEEF to the JDK source.
> 
> I was initially of the opinion that the solution already provided here was enough. We could potentially have added a Git hash in addition to the version/tag to more precisely and permanently identify the exact Sleef source we are depending on. At least a Git hash wouldn't change externally.
> 
> However, Andrew's arguments have swayed me. I now think we should add a more or less complete dump of the Sleef source into the JDK repository. I'm still open to trimming it down somewhat as long as the build steps we need to run to generate the headers we need, using the Sleef build system, are still functional. I'm basically agreeing with his suggestion but will spell it out in detail here for completeness in this PR.

Sounds good, thanks.

> We should then add a script that automatically performs the necessary build steps, using the Sleef official build system, to generate the headers we need, and outputs them into the JDK source tree, in the location where we will also commit those headers. This script should document what dependencies and configuration is necessary to run it, which will include cmake and possibly other things. Performing this step doesn't need to be completely streamlined, just reasonably possible to run. It's meant to be an import/verification step. With this solution I would recommend putting the script next to the Sleef source tree instead of in make/devkit/.
> 
> The normal JDK build will just use the committed pre-generated headers.
> 
> I looked briefly at the heroic build work provided by @fitzsim and while I admire the effort, I don't think this is the right way and we already dismissed this approach earlier. Not because I didn't think it was feasible to implement, but because of the future maintenance burden. The Sleef build is non trivial so we shouldn't try to replicate it in our build. The risk of our implementation diverging in the future is too great.

OK.

I had a brief meeting with Mark Reinhold to discuss this issue. I asked him if it would ever be acceptable to check in preprocessed code without the corresponding preferred from, and the immediate answer was an emphatic "no". So that's that.

I'm neutral on just how much of SLEEF should be checked into our source tree. As long as we include the full source for whatever subset of SLEEF we use, and it is the real source in its preferred form, I'm happy.

**Comment by vidmik** [comment]
I'm following up on the necessary logistics given the new plan.

**Comment by vidmik** [comment]
@Hamlin-Li We now have the necessary approvals in place for the new plan to include all of SLEEF (and the pre-generated header files) in the JDK. I (or somebody else from Oracle) will need to be the one committing/contributing the actual SLEEF code itself in the end. Can you take on the work of implementing the actual plan/logic and make sure it's all effectively ready for integration?

**Comment by Hamlin-Li** [comment]
@vidmik Thanks for pushing forward this work.
Sure, I can take the work.

Just several questions, Sorry, as this is the first time I do this kind of work in jdk.
In this pr, we already have the pre-generated sleef files and the scripts to generate it, what we are missing is the sleef original file (in my understanding, we also need to checkin them), besides of sleef original files, do we need any other files or document? In particular, do you mind to clarify further about `plan/logic` mentioned above?

**Comment by vidmik** [comment]
@Hamlin-Li Thank you for taking it on!

Part of the work is figuring out exactly what exactly we want - see the comments from @erikj79 and @theRealAph in this PR for inspiration. In particular, we'll want to include all or part of the SLEEF sources in the JDK, but we'll likely also want some additional build scripts and documentation to make it easy for developers to reproduce and update going forward. This could probably be based on the README and createSleef script.

**Comment by Hamlin-Li** [comment]
@vidmik Thanks for clarifying, I'll figure it out.

**Comment by Hamlin-Li** [comment]
By understanding the previous discussion, there should be 2 parts of source code related to sleef integrated into jdk: 
1]. original sleef source
2]. pre-processed sleef header files

Besides of above sources, there should be another part connecting above 2 parts, i.e. 
3]. scripts to transfering from 1] to 2], and necessary documents, e.g. record which tag of sleef to use, .


So, for the source layout/organization, the previous conclusion is to have both 1] and 2], and 1] is just for verification and reproducing, general build of jdk only needs 2].

As to where to put various new files, I think current location for 2] in this pr is good, 1] could be put somewhere with minimum impact on current build system, `src/utils/sleef` suggested by @magicus is a good candidate, 3] should be put together with 1].
In summary, one solution could be as below:
```
${JDK_ROOT}/src
  |-- utils/
      |-- sleef-src/   # i.e. part 1]
           |-- ... # original sleef source, we could trim or not trim the src tree. 
      |-- sleef-misc/
          |-- createSleef.sh    # accept a sleef source root(i.e. part 1], or an external sleef repo location) as input, and generate sleef header files(i.e. part 2] )
          |-- sleef.md
          |-- ..  # other misc files, e.g. 
  |-- jdk.incubator.vector/linux/native/libvectormath      # i.e. part 2]
      |-- README  # also including a reference pointing to src/utils/sleef*
      |-- misc.h
      |-- sleefinline_advsimd.h
      |-- sleefinline_rvvm1.h
      |-- sleefinline_sve.h
```

As for which part of original sleef source to integreate into jdk, we could trim the original source in some way, but I think it's more simple and maintainable to just put all sleef source into jdk without any trim.

**Comment by erikj79** [comment]
> 3]. scripts to transfering from 1] to 2], and necessary documents, e.g. record which tag of sleef to use, .

In addition to the tag, we should include the full git hash as well. A tag can be moved, but it's very hard to fake a new git hash for a commit if any changes are made.

**Comment by Hamlin-Li** [comment]
> > 3]. scripts to transfering from 1] to 2], and necessary documents, e.g. record which tag of sleef to use, .
> 
> In addition to the tag, we should include the full git hash as well. A tag can be moved, but it's very hard to fake a new git hash for a commit if any changes are made.

I agree, thanks for reminding! I also updated my comment above.

**Comment by magicus** [comment]
After thinking a bit more on this, I would recommend a somewhat different layout:

```
${JDK_ROOT}/src
  |-- jdk.incubator.vector/linux/native/libsleef
      |-- README.md
      |-- upstream/ 
           |-- ... # original sleef source (trimmed or not)
      |-- generated/
           |-- misc.h
           |-- sleefinline_advsimd.h
           |-- sleefinline_rvvm1.h
           |-- sleefinline_sve.h
  |-- jdk.incubator.vector/linux/legal/libsleef.md
```

`jdk.incubator.vector/linux/legal/libsleef.md` needs to contain the libsleef
license; see other legal/*.md files for reference.

The README.md file should briefly but clearly explain how we handle libsleef
code in the JDK, explain how to update the generated files, and describe
from where the source was obtained and at what git tag+hash.

**Comment by magicus** [comment]
In the build system, we will then add jdk.incubator.vector/linux/native/libsleef/generated as an additional source root when building libvectormath.

**Comment by magicus** [comment]
Furthermore, despite what Erik said above, I would really, really like to not have a stand-alone shell script mixed in with the spleef sources. Instead, we should treat updating the generated spleef sources as any other build-related activities, that is, we should have a build target for that.

I imagine we would have a `make update-spleef-source` that calls into UpdateSpleefSource.gmk. This in turn will more or less call `cmake` with the proper arguments needed to build spleef, and copy the resulting files into the `generated` directory.

The prerequisites required for doing this, e.g. having cmake available (and possibly any other requirements?) should be checked by configure. 

There are two slightly different nuances on how we handle these dependencies: We could either add a `--enable-spleef-updating` argument to configure, which will fail configure if some dependencies are missing, and if will be responsible for making the `update-spleef-source` target available. (This would have a typical true/false/auto setup, so the target will be available if you happen to have all requirements present.)

The other alternative, if the one above would be too much, is to just set the dependencies in configure if they are found, and then verify in UpdateSpleefSource.gmk that they are non-empty (and fail with an error otherwise). This is contrary to the general design of the build system, but might be acceptable in a special case like this, to gain a bit of simplicity in the code.

**Comment by magicus** [comment]
I can help you with the build system changes, if what I wrote above made no sense to you.

**Comment by erikj79** [comment]
> Furthermore, despite what Erik said above, I would really, really like to not have a stand-alone shell script mixed in with the spleef sources. Instead, we should treat updating the generated spleef sources as any other build-related activities, that is, we should have a build target for that.

I'm ok with this as you are offering to help implement it. My suggestion for using a separate script was out of simplicity and an assumed lack of resourcing for a more involved solution. Your suggested solution would generally be a better end result.

**Comment by Hamlin-Li** [comment]
> ```
> ${JDK_ROOT}/src
>   |-- jdk.incubator.vector/linux/native/libsleef
>       |-- README.md
>       |-- upstream/ 
>            |-- ... # original sleef source (trimmed or not)
>       |-- generated/
>            |-- misc.h
>            |-- sleefinline_advsimd.h
>            |-- sleefinline_rvvm1.h
>            |-- sleefinline_sve.h
>   |-- jdk.incubator.vector/linux/legal/libsleef.md
> ```

Thanks, this is much better.

> Furthermore, despite what Erik said above, I would really, really like to not have a stand-alone shell script mixed in with the spleef sources. Instead, we should treat updating the generated spleef sources as any other build-related activities, that is, we should have a build target for that.
> ...
> There are two slightly different nuances on how we handle these dependencies.

I'm ok too with either of solutions.
Only reminding is that we should treat sleef build/source-generation as a blackbox.

**Comment by magicus** [comment]
@vidmik @Hamlin-Li I kind of hijacked this... 

I created the necessary build system changes to update the generated sleef code. At that point, also importing the actual source code was just a trivial addition.

The new PR is here: https://github.com/openjdk/jdk/pull/20781

I think it implements everything we've talked about here, in the way I understood consensus to be.

**Pull request closed without merging**
